File: app/src/main/java/com/kaciula/archiman/infrastructure/ArchimanApplication.java
package com.kaciula.archiman.infrastructure;

import com.kaciula.archiman.BuildConfig;
import com.kaciula.archiman.domain.usecases.InitColdStart;
import com.kaciula.archiman.injection.AppComponent;
import com.kaciula.archiman.injection.AppModule;
import com.kaciula.archiman.injection.DaggerAppComponent;
import io.realm.Realm;
import io.realm.RealmConfiguration;
import javax.inject.Inject;

public class ArchimanApplication extends BaseApplication {

  private AppComponent appComponent;

  @Inject InitColdStart initColdStart;

  @Override
  public void onCreate() {
    super.onCreate();

    appComponent = DaggerAppComponent.builder().appModule(new AppModule()).build();
    appComponent.inject(this);

    setupRealm();

    initColdStart
        .execute(new InitColdStart.RequestModel(BuildConfig.VERSION_CODE))
        .subscribe();
  }

  public static AppComponent component() {
    return ((ArchimanApplication) getContext()).appComponent;
  }

  public static ArchimanApplication get() {
    return (ArchimanApplication) getContext();
  }

  public boolean isCrashlyticsUsed() {
    return false;
  }

  private void setupRealm() {
    Realm.init(this);
    RealmConfiguration realmConfiguration = new RealmConfiguration.Builder().build();
    Realm.setDefaultConfiguration(realmConfiguration);
  }
}

File: app/src/main/java/com/kaciula/archiman/infrastructure/ArchimanApplication.kt
package com.kaciula.archiman.infrastructure

import com.kaciula.archiman.BuildConfig
import com.kaciula.archiman.domain.boundary.infrastructure.CrashReporter
import com.kaciula.archiman.domain.usecases.InitColdStart
import com.kaciula.archiman.domain.util.Timberific
import com.kaciula.archiman.injection.AppComponent
import com.kaciula.archiman.injection.AppModule
import com.kaciula.archiman.injection.DaggerAppComponent
import io.realm.Realm
import io.realm.RealmConfiguration
import javax.inject.Inject

abstract class ArchimanApplication : BaseApplication() {

    private lateinit var appComponent: AppComponent

    @Inject lateinit var initColdStart: InitColdStart
    @Inject lateinit var crashReporter: CrashReporter

    override fun onCreate() {
        super.onCreate()

        appComponent = DaggerAppComponent.builder().appModule(AppModule()).build()
        appComponent.inject(this)

        Timberific.init(BuildConfig.DEBUG)

        setupRealm()

        initColdStart
                .execute(InitColdStart.RequestModel(BuildConfig.VERSION_CODE))
                .subscribe()
    }

    open fun isCrashlyticsUsed() = false

    private fun setupRealm() {
        Realm.init(this)
        Realm.setDefaultConfiguration(RealmConfiguration.Builder().build())
    }

    companion object {

        fun component(): AppComponent {
            return get().appComponent
        }

        fun get(): ArchimanApplication {
            return context as ArchimanApplication
        }
    }
}
